### PR TITLE
overloaded app::install() that uses a callback to get data

### PR DESCRIPTION
**Pull request by ksanislo**
This is my proposed extension to allow installing from a variety of data sources by use of a callback. This keeps all the messy compression/encryption code out of libcitrus, while still allowing for direct stream installs from pretty much any data source someone can come up with.

### CODE REVIEW

File: source/citrus/app.cpp [hunk 1]
```diff
@@ -385,6 +386,65 @@ void ctr::app::install(ctr::fs::MediaType mediaType, FILE* fd, u64 size, std::fu
...
+    };
+    memset(buf, 0, bufSize);
+
+    Handle ciaHandle;
+    ctr::err::parse(ctr::err::SOURCE_AM_START_CIA_INSTALL, (u32) AM_StartCiaInstall(mediaType, &ciaHandle));
+    if(ctr::err::has()) {
+        return;
+    }
+
+    ret=(s32)HTTPC_RESULTCODE_DOWNLOADPENDING;
```
Comment (Steveice10) [inline source/citrus/app.cpp]: Probably shouldn't be HTTPC related if this is now a generic callback-based function?

**Comment by ksanislo** [comment]
I've removed the HTTPC specifics, now the callback returns int < 0 to signal more data, and >=0 to signal the end of the file. This was chosen because HTTPC returns -666853333 natively for pending data (allowing super hacky passthrough), and 0 and >= 1 can still mean success and abnormal exit for the callback, both of which will signal the end to the loop.

**Comment by Steveice10** [comment]
Kind of merged that weird, but whatever.

**Comment by ksanislo** [comment]
Sorry about that, I've still not figured out how to `git` properly.

**Pull request merged**